Add bitwise or support

sethterashima opened the pull request:
Allows bitwise OR (`|`) operations in Kafel policies.

Adds test harness support for system calls that return a file descriptor (new `SYSCALL_RESULT_HANDLE_SPEC` macro).

This commit resolves #14.

sethterashima commented:
I signed it!

sethterashima commented:
> Generally looks good, but needs either:
> 
> * some more codegen modifications to use bpf M[] as stack to evaluate more complex expressions + compile time error when out of stack
> * just a compile error when for expressions which would require the stack

This fell off my plate for a while, but I went with the first option. I included code to determine if the stack is required in order to avoid the overhead of unnecessary push/pop operations, but the logic for that should have some careful review. I've added some tests and verified that the BPF code for the profiles in `samples` isn't changed by this commit.

happyCoder92 commented on test/basic.c:
```diff
@@ -102,3 +103,54 @@ TEST_CASE(rules_order) {
+       POLICY a { \n\
+         ALLOW { exit, close, open { flags == O_WRONLY|O_CREAT }} \n\
+       } USE a DEFAULT KILL");
+  TEST_POLICY_ALLOWS_SYSCALL(SYSCALL_SPEC3(__NR_open, (long) "test_file", O_WRONLY|O_CREAT, 0600),
```
this makes the test not fully isolated (e.g. there is "test_file" in cwd or cwd is not writable)
Use `__NR_getpid` and:
`getpid(arg) { arg == O_WRONLY|O_CREAT) }`
instead. Also no need to change harness and Makefile then.

sethterashima replied:
Switched to `read`. The test cases don't make much sense semantically since they operate on different parameters, but this is required to exercise code outside of the expression tree simplification step (which is why I missed the clobbered X-register issue).

happyCoder92 commented on src/codegen.c:
```diff
@@ -312,10 +323,26 @@ static bool is_const_value(struct expr_tree *expr, int word) {
+// Returns 1 if we need to push the result of executing the right sub-tree onto
+// the stack (as opposed to the index register) before evaluating the left
+// sub-tree.
+static int should_use_stack(struct expr_tree *left, struct expr_tree *right, int word) {
```
use `bool`

sethterashima commented:
Thanks for the review. I'll pick this up when I get back from vacation next week.

sethterashima commented on src/codegen.c:
```diff
@@ -298,6 +300,15 @@ static bool is_const_value(struct expr_tree *expr, int word) {
       return true;
     case EXPR_VAR:
       return false;
+    case EXPR_BIT_OR:
```
Expression tree is now preprocessed to cache these values and store them in the tree structure.

sethterashima commented:
Addressed the above issues and added another commit that does some minor optimization of the emitted code. Let me know if you want the optimization to be a separate PR; I thought it was minor enough to stick in since I was refactoring the affected code anyway.

happyCoder92 commented on src/codegen.c:
```diff
@@ -280,77 +282,107 @@ static int add_jump_set(struct codegen_ctxt *ctxt, __u32 what, int tloc,
+      return;
     default:
-      ASSERT(0);  // should not happen
+      cache_constants_by_word(expr->right, word);
```
`ASSERT(expr->type >= EXPR_BINARY_MIN && expr->type <= EXPR_BINARY_MAX)` just in case.

sethterashima commented:
@happyCoder92 everything look good?

happyCoder92 commented:
Sorry, for the delay, didn't notice you updated the PR addressing the last comments.
Thanks for working on this one, awesome job!
